### PR TITLE
Additional support for RAG hyperparameter optimization.

### PR DESCRIPTION
**Pull request by acompa**
Introducing the hyperparameter optimization logic to auto-optimize user RAG workflows.

Note that this PR also glues together various elements: workflow definition, evaluation, test dataset logic. Let's not merge this directly into `main`, but into #49 instead.

### CODE REVIEW

File: tests/test_toolkit/rag/test_example_workflow.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                        "openai_small",
+                        "openai_large",
+                        "cohere_eng",
+                        "cohere_multi",
+                    ]
+                ),
+            ],
+            "BM25",
+        ),
+        "temperature": (Annotated[float, FloatValueRange(0.0, 1.0, 0.1)], 0.9),
```
Comment (r0ymanesco) [inline tests/test_toolkit/rag/test_example_workflow.py:46]: This param is not used in the `rag_workflow`. Just want to make sure that `FloatValueRange` is working correctly

File: tests/test_toolkit/rag/test_example_workflow.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    IntValueRange,
+)
+
+
+class ExampleNDRagWorkflow(BaseNDRagWorkflow):
+    parameter_specs = {
+        "chunk_size": (Annotated[int, IntValueRange(1000, 2500, 500)], 1000),
+        "chunk_overlap": (Annotated[int, IntValueRange(50, 200, 25)], 100),
+        "top_k": (Annotated[int, IntValueRange(1, 20, 1)], 5),
+        "algo": (
```
Comment (r0ymanesco) [inline tests/test_toolkit/rag/test_example_workflow.py:31]: This param is not used in the `rag_workflow`. Just want to make sure that `CategoricalValueOptions` is working correctly

**Review by r0ymanesco — APPROVED** [review]
Approving but would be great if you can address those 2 minor comments before merging.

File: tests/test_toolkit/rag/test_example_workflow.py (hunk 1 above)
Comment (acompa) [reply tests/test_toolkit/rag/test_example_workflow.py:46]: Funny enough another test surfaced an issue with this `FloatValueRange` - while it is not accessed in `rag_workflow` we do set and retrieve values during `auto_optimize`. That said I've added a `print` to resolve it.

Comment (r0ymanesco) [reply tests/test_toolkit/rag/test_example_workflow.py:46]: Ok great thanks, feel free to merge

**Pull request merged**